### ISSUE TITLE
Idea for evolving split repos around client, apimachinery, etc.

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#41629]
@lavalamp  and I spoke for a couple hours today, and I need to write some thoughts down.  Maybe it will go nowhere, but let's see.

We have a small set of things that are/were in the main repo which really MAKE SENSE outside of it.  The flagship of this is `client-go`, which I would expect to have orders of magnitude more consumers than anything else in the project.  It just makes sense to make that easier to consume.  Other examples that make sense are the machinery stuff, utils, generic API server.  This is a distinctly different set of things than kubelet, kube-proxy, etc., and I think we might benefit from some "arms length" relationship management.

Several of these are sort of half-extracted already, but what we have today is a weird in-between thing that is utterly incomprehensible.  I think we either need to push ahead or roll back to a mega-mono-repo until we have a better clue what to do about it.

So here's the idea (NB, this is not a proposal yet, just an idea).  We bust out a few new repos that operate on their own.  Development is done IN THESE REPOS, not in the "main" repo.  No collective e2e gates, just good old fashioned repos with CI and unit tests and things like that.

k8s.io/api - all of our versioned API groups and types.  None of the code, just the "spec" which we happen to express in our custom IDL, which just happens to be syntactically identical to Go structs.  The list of repo owners and approvers is very small and PRs against this repo can be watched carefully.  TBD how to manage the current codegen stuff, but I think we can make it work.  There's no "release" per se on this, but it would be tagged alongside kube releases.

k8s.io/lib - all of our *really* generic utility libs.  Carefully curated and not allowed to import anything else in kubernetes.  Limited external deps.  anything in kube can depend on this, including api (for things like IntOrString, sigh).  Treat this as Go expects libs to operate - head always works.  Maybe version it through something like gopkg?

k8s.io/client-go - the Go client lib.  consumes lib and api.  Is consumed by the "main" repo and all the shard repos that feed into our system.  Eat your dogfood.

k8s.io/api_machinery - all of the things we need to make our API semantics happen. In particular, some of the serialization stuff is needed by the client-go

k8s.io/api_server - the generic API server framework.  Consumed by heapster, kube-apiserver, etc.  Maybe this can be part of api_machinery?

(names can be tweaked :)

We treat these things as if a third-party owned them.  The main repo e2e tests will still catch failures, but only when we vendor these in, just like all of our other external deps.  If this turns out to ACTUALLY be detrimental, we can explore shared CI.

Yes, this means a change to (e.g.) client has to wait for it to be re-vendored.  If we touch vendoring enough we can make it easier or more automatic to update.  Likewise an API change becomes more complicated - change api repo, merge, vendor to main repo, update internal type and conversions and regen everything, etc, merge.

Get rid of all the `staging` crap we have today.

Cut the weird spaghetti deps between client stuff and the main repo - have to use versioned types EVERYWHERE, with the exception of kube-apiserver (and eventually maybe not even that).

I feel like this is a ton of work, but tractable.  It doesn't depend on technology that doesn't exist yet.  It isn't inventing new processes - it is the devil we already know.

We can fight about if/how to split development of kubelet from scheduler from controller manager in another thread.

Thoughts?

@philips @deads2k @lavalamp @caesarxuchao @mbohlool @bgrant0607 @brendandburns @smarterclayton

**Comment by deads2k** [comment]
I'm generally in favor of the idea, but I'm not sure you've made the repercussions sound scary enough.

Having worked downstream of a large project for a while, I'd be concerned about having uncoordinated updates to a repo that k8s.io/kubernetes depends upon.  You can very easily end up in a state where you have a change in one of the precursor repos that are very difficult to be consumed by the k8s.io/kubernetes repo.  It's also easy to end up "half-pregnant" with half of your repos on new levels of X and half on new levels of Y and the work to reconcile it is massive.  Even people very familiar with the overall project are likely to have difficulty predicting and handling changes since you'll probably have multiple changes in each vendoring.

Adding other OpenShift rebasers: @liggitt @ncdc @sttts @smarterclayton

**Comment by deads2k** [comment]
> k8s.io/api_server - the generic API server framework.  Consumed by heapster, kube-apiserver, etc. Maybe this can be part of api_machinery?

I don't think we should rope this in with apimachinery.  The two have different concerns and only touch in a couple interfaces, strictly one way.

**Comment by smarterclayton** [comment]
This would probably be yet another thing a new contributor (or even moderate ones) would run up against as a huge roadblock (revendoring and dealing with rebase is ugly) and get angry about.

Having a compact and low dep count library that is stable seems goal #1.  Is making it easy to submit small patches to that library as important?

Having the subdivided repos inside staging is growing on me, but even the subdivided set doesn't seem like it solves goal 1 without a transform step.  I'm still a bit more leaning towards client-go as an output, not an input, just to be able to evolve it on a different rate than the core work.

**Comment by bgrant0607** [comment]
ref #24343

**Comment by sttts** [comment]
Like @deads2k I have serious doubts that late integration via on-demand vendoring is helping us. This has caused so much pain in the OpenShift rebases that everybody involved there will have a bad vibe with the idea to do the same upstream.

@thockin I would like to understand what you say about the staging dir: "a weird in-between thing that is utterly incomprehensible".

I understand that we had some tooling issues, e.g. with codegen and bazel ignoring the tests. Moreover, that symlink in `vendor/` is a bit "non-standard", call it weird if you want. But it seems to work surprisingly well with the go toolchain.

But other than that: what exactly is incomprehensible compared to real repositories? Certainly there is a learning curve to know what is where and where to put new code. But this is exactly the same if we move to real repos.

As @smarterclayton wrote, staging has grown also on me over time. We don't have the huge tooling issues discussed in https://github.com/kubernetes/kubernetes/issues/24343. At the same time, we enforce layering of our code pretty effectively with the `verify-staging-imports.sh` script, in my eyes (next to the Github dev process scaling issues) the main advantage we expected from moving to real independent repos.

**Comment by sttts** [comment]
Following @thockin list of repos, I propose to implement some of them as further staging directories (to start with something before the tooling issue is solved): 
1. create a stable `k8s.io/lib` or `k8s.io/foundation` in staging
   - move relevant `pkg/utils` from apimachinery, client-go and kube there.
   - @thockin volunteered as a gate keeper 👍  
2. move all types into `k8s.io/api` in staging
   - as minimal as possible. Probably including deepcopy, defaulting, conversion. Validation to be discussed.
   - make kube **and** client-go depend on that, i.e. remove our insane type duplication in `pkg/api` and `k8s.io/client.go/pkg/api`
   - while splitting out the core group from `pkg/api` into `k8s.io/api/core`.
3. remove generated clients and informers in kube completely. Instead, generate into client-go (in staging) directly.

Especially, **(2) should be done as soon as possible**. It's purely mechanical, can be done probably in a day or two. Removes a lot of friction from people who have not the background knowledge about the latest crazy copy.sh magic client-go is doing. This should be done **before client-go 2.0-GA or 2.0+1 is tagged**.

Something like https://github.com/kubernetes/kubernetes/pull/41689 can be implemented to make the multi repo approach bearable for client-go users.

**Comment by deads2k** [comment]
@smarterclayton had originally convinced me that we should make copies of API objects to make it more reasonable to get generate clients with old API guarantees against new objects, but if we get https://github.com/kubernetes/kubernetes/pull/41486, we should be able to regenerate clients whenever we need to.

I'm ok with splitting the API objects into a separate staging dir and eliminating the only true duplication in the codebase we have at the moment.  We should try to decide on Monday, to make the pulls and deal with any fallout before code freeze.

**Comment by thockin** [comment]
@sttts 
> This has caused so much pain in the OpenShift rebases 

Note that I am exclusively talking about things that are really distinct libraries - NOT about core components of the system.  Vendoring today kind of sucks, but it is getting better and it is a well-understood and totally unsurprising form of suck.  I'd like a little less magic in our tree...

@sttts 
> I would like to understand what you say about the staging dir: "a weird in-between thing that is utterly incomprehensible".

* Nobody who knows anything about go knows WTF is going on
* I don't know how to regen or refresh or whatever I need to do to it
* I don't know what code I am supposed to edit vs not
* The symlinks in vendor/ means all sorts of go tools are probably subtly broken - Go *does not* support symlinks and is known to barf in cases you would expect to just work.  We have to wrap standard tools to remove and recreate symlinks properly.
* Having multiple import roots in the same repo has already caused problems for the build system, and is AT BEST confusing for codegen
* It doesn't actually solve any of the problems of being able to delegate things like label and issue management, ownership rights, etc.  This might not matter much NOW but the cast of characters WILL change.
* In short, it flies in the face of any established practice and it confuses and confounds newcomers and longtimers alike.

@sttts 
> Especially, (2) should be done as soon as possible.

PLEASE do not do any more staging things until we have some sort of agreement on direction here.

**Comment by smarterclayton** [comment]
I'm having trouble processing your concerns, mostly because I haven't seen any of the problems you've brought up yet directly.

Regen or refresh: run make like normal
Which files to edit: there's only one file to edit
Build system problems: which ones in particular?  Maybe I missed some recently, but everything was ticking over normally last week.
* confuses and confounds: this is somewhat hyperbolic, but the initial set of make changes, bazel, and changing builds to be in docker were all confusing for people at first.  We got over it.  Nothing here is any more impactful.  I think crying foul here is a bit of a stretch.

**Comment by thockin** [comment]
On Sat, Feb 18, 2017 at 5:02 PM, Clayton Coleman
<notifications@github.com> wrote:
>
> I'm having trouble processing your concerns, mostly because I haven't seen any of the problems you've brought up yet directly.
>
> Regen or refresh: run make like normal

Are you saying there are no dup files anywhere any more?

> Which files to edit: there's only one file to edit

Several times I have experience weirdness like a bulk search/replace
that fails until I revert staging/...

> Build system problems: which ones in particular? Maybe I missed some recently, but everything was ticking over normally last week.

Well, for example, any `make` that takes a WHAT arg was broken by
changes related to staging.  No fix yet, and I am not sure it's
fixable.  It's relatively minor, but it's just one that popped up and
surprised me.

It's a totally legitimate thing for go tools to expect that a single
repo, with the exception of vendor/, holds a single go import tree,
and I have been repeatedly burned by trying to do things with symlinks
in the path to packages.  How many bits of our tools work by
prepending `k8s.io/kubermetes` to a relative path?  This, to me, is
the biggest issues.  It's just so wildly divergent from what anyone
else is doing that it's bound to break.

> confuses and confounds: this is somewhat hyperbolic, but the initial set of make changes, bazel, and changing builds to be in docker were all confusing for people at first. We got over it. Nothing here is any more impactful. I think crying foul here is a bit of a stretch.

Nobody but a small set of people is expected to ever touch bazel
except in very rare cases (it's not done yet, obviously).  `make` is
the universal standard interface to compiling code, and the fact that
`go build` doesn't work is still a pain point.   Claiming other bad
decisions were made does not justify more bad decisions.

If we were starting over today, what would we do?  I doubt it's this....

**Comment by smarterclayton** [comment]
> Are you saying there are no dup files anywhere any more?

All the dup files are slimmed down dependencies caused by stuff checked in.  All those packages should have comments saying "this has been deprecated and shouldn't be changed because we were dumb and allowed circular dependencies".  They also should have been marked readonly which last I checked was gated by verify checks.

> Several times I have experience weirdness like a bulk search/replace that fails until I revert staging/...

Ok, had not seen anything like that.  The last bit that @sttts was talking about was client-go - that's the one that has been that way for a very long time, and needs to be made consistent.

> Well, for example, any `make` that takes a WHAT arg was broken by changes related to staging.  No fix yet, and I am not sure it's fixable.  It's relatively minor, but it's just one that popped up and surprised me.

Sure - I'm not saying there have been no problems, just that previous massive changes people have made to the system occasionally miss things.  I have a hard time seeing that as the price of change, and certainly we've treated it like that previously.

> It's a totally legitimate thing for go tools to expect that a single repo, with the exception of vendor/, holds a single go import tree, and I have been repeatedly burned by trying to do things with symlinks
in the path to packages.  How many bits of our tools work by prepending `k8s.io/kubermetes` to a relative path?  This, to me, is the biggest issues.  It's just so wildly divergent from what anyone else is doing that it's bound to break.

I'm going to be honest.  At this point what Go tools "expect" is part of the problem.  I've sorted through lots of parts of this issue, and I've hit a few assumptions, but the problems haven't been the tools, it's been the rest of the stack (make, generation, vendored dependencies).  That complexity isn't going away, because we have that complexity.

> Nobody but a small set of people is expected to ever touch bazel except in very rare cases (it's not done yet, obviously).  `make` is the universal standard interface to compiling code, and the fact that
`go build` doesn't work is still a pain point.   Claiming other bad decisions were made does not justify more bad decisions.

Are you saying that you think that bazel and make were bad decisions?  I was saying that we occasionally make changes that introduce more complexity because we hope there's a payoff.  We don't always reap the benefits, and other folks on the team are impacted who were not directly involved in the decisions.  The previous state of staging, client-go, our dependency tree, and the interdependencies between library functions was unacceptable and causing more impact to the folks trying to clean these problems up than was allowable to continue.  Those people made a series of decisions that in their best judgement would push the state forward, and got most things right, a few things were broken, and some new complications cropped up in the system.  My opinion is that the impact to others was less than those other changes that I mentioned, and the benefits are more significant moving forward.

> If we were starting over today, what would we do?  I doubt it's this....

The more I think about, the more I think it would be.  Having nested staged repos that are copied out preserves most of the benefits of the monorepo.  The one thing I would probably change would be to create a sub GOPATH like etcd has that constructs a real tree for packages and then build inside it.  However, throwing up our hands and doing nothing I don't consider a real option, and I do feel that the folks working on this have tried to minimize the impact to the degree that everyone else making important changes has done.

It sounds like there's a discussion happening anyway - we can talk more about this then.

**Comment by thockin** [comment]
Mobile now.

Your explanations make some sense, but it still feels very weird and
backwards.

> At this point what Go tools "expect" is part of the problem

Maybe but I don't think we should pretend it isn't real.  If we could get
the Go team to at least commit to not breaking in the face of symlink, I
would feel better.

None of this addresses the limitations of GitHub.  I still can't filter
well.  Our bots are very custom and not always obvious, and they don't
handle labels.

This is as much a problem as the rest, isn't it?


On Feb 18, 2017 5:52 PM, "Clayton Coleman" <notifications@github.com> wrote:

Are you saying there are no dup files anywhere any more?

All the dup files are slimmed down dependencies caused by stuff checked in.
All those packages should have comments saying "this has been deprecated
and shouldn't be changed because we were dumb and allowed circular
dependencies". They also should have been marked readonly which last I
checked was gated by verify checks.

Several times I have experience weirdness like a bulk search/replace that
fails until I revert staging/...

Ok, had not seen anything like that. The last bit that @sttts
<https://github.com/sttts> was talking about was client-go - that's the one
that has been that way for a very long time, and needs to be made
consistent.

Well, for example, any make that takes a WHAT arg was broken by changes
related to staging. No fix yet, and I am not sure it's fixable. It's
relatively minor, but it's just one that popped up and surprised me.

Sure - I'm not saying there have been no problems, just that previous
massive changes people have made to the system occasionally miss things. I
have a hard time seeing that as the price of change, and certainly we've
treated it like that previously.

It's a totally legitimate thing for go tools to expect that a single repo,
with the exception of vendor/, holds a single go import tree, and I have
been repeatedly burned by trying to do things with symlinks
in the path to packages. How many bits of our tools work by prepending
k8s.io/kubermetes to a relative path? This, to me, is the biggest issues.
It's just so wildly divergent from what anyone else is doing that it's
bound to break.

I'm going to be honest. At this point what Go tools "expect" is part of the
problem. I've sorted through lots of parts of this issue, and I've hit a
few assumptions, but the problems haven't been the tools, it's been the
rest of the stack (make, generation, vendored dependencies). That
complexity isn't going away, because we have that complexity.

Nobody but a small set of people is expected to ever touch bazel except in
very rare cases (it's not done yet, obviously). make is the universal
standard interface to compiling code, and the fact that
go build doesn't work is still a pain point. Claiming other bad decisions
were made does not justify more bad decisions.

Are you saying that you think that bazel and make were bad decisions? I was
saying that we occasionally make changes that introduce more complexity
because we hope there's a payoff. We don't always reap the benefits, and
other folks on the team are impacted who were not directly involved in the
decisions. The previous state of staging, client-go, our dependency tree,
and the interdependencies between library functions was unacceptable and
causing more impact to the folks trying to clean these problems up than was
allowable to continue. Those people made a series of decisions that in
their best judgement would push the state forward, and got most things
right, a few things were broken, and some new complications cropped up in
the system. My opinion is that the impact to others was less than those
other changes that I mentioned, and the benefits are more significant
moving forward.

If we were starting over today, what would we do? I doubt it's this....

The more I think about, the more I think it would be. Having nested staged
repos that are copied out preserves most of the benefits of the monorepo.
The one thing I would probably change would be to create a sub GOPATH like
etcd has that constructs a real tree for packages and then build inside it.
However, throwing up our hands and doing nothing I don't consider a real
option, and I do feel that the folks working on this have tried to minimize
the impact to the degree that everyone else making important changes has
done.

It sounds like there's a discussion happening anyway - we can talk more
about this then.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/41629#issuecomment-280889429>,
or mute the thread

**Comment by sttts** [comment]
>
>
>    - I don't know how to regen or refresh or whatever I need to do to it.
>    I don't know what code I am supposed to edit vs not.
>
> The last non-authorative code that is a copy is `
k8s.io/client-go/pkg/api*`. Everything else can be edited (now, not 3
months ago when client-go was a 90% copy). That's why I propose k8s.io/api
to get this out of the way finally. No more "what can I edit and how can I
sync".

>
>    - The symlinks in vendor/ means all sorts of go tools are probably
>    subtly broken - Go *does not* support symlinks and is known to barf in
>    cases you would expect to just work. We have to wrap standard tools to
>    remove and recreate symlinks properly.
>
> Technically there is absolute no reason to have staging/ at all (with
k8s.io/api even staging/copy.sh would go away). We could put the code
directly into the vendor dir. In the worst case, your vendoring tool of
choice will fall over it. If the symlink is really such an issue, this
might be an option. Having authorative code in vendor/ is weird though as
well.

**Comment by smarterclayton** [comment]
> Maybe but I don't think we should pretend it isn't real.  If we could get the Go team to at least commit to not breaking in the face of symlink, I would feel better.

Yeah, not arguing against resolving.  Just trying to emphasize that we have a little meta-island of stability that I think is net better than the split repos right now.  I'm ok with not being super aggressive on new repos in staging (in fact I'd agree that we should have just a few), but I think it's an ok point to be for right now that is a halfway point.  And I haven't seen the tools blow up, I've seen our tools blow up.

> None of this addresses the limitations of GitHub.  I still can't filter well.  Our bots are very custom and not always obvious, and they don't handle labels.

Yes, but I would expect the OWNERS / issues lists for all of the current staging repos to be sig/api-machinery and fairly limited.  Client-go is weird, because any bug there is likely a bug in the main repo, and if we break the generation pattern by accepting downstream fixes then we might as well just cast it adrift right now.  I would definitely like to see something better here, no disagreement.  But they still feel very sig managed and constrained.

**Comment by caesarxuchao** [comment]
> The last non-authorative code that is a copy is `k8s.io/client-go/pkg/api*`.

To be precise, in client-go, apart from the APIs, [pkg/util and pkg/version](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/client-go/pkg) are also copied, but I think we *can* make them authoritative.

**Comment by lavalamp** [comment]
> move all types into k8s.io/api in staging

I can't keep up with all the above discussion but I have very strong opinions about what should be published in such a repo--please don't merge such a change without letting me take a look at it.

**Comment by caesarxuchao** [comment]
Another inconvenience caused by `staging/`: it breaks code navigation. I'm using vim-go, which can't jump to definitions in `staging/`. Has anyone figured out a workaround other then `grep`?

cc @Random-Liu

**Comment by ncdc** [comment]
@caesarxuchao I use vim-go and am able to jump to definitions in staging

**Comment by caesarxuchao** [comment]
@ncdc does it take you to staging or to the copy in GOPATH? Do you have to do any tricks in GOPATH?

**Comment by ncdc** [comment]
@caesarxuchao it takes me into the vendor directory, e.g. `~/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/interfaces.go`. And `vendor/k8s.io/apimachinery` is just a symlink to inside staging.

My GOPATH is just `$HOME/go`

**Comment by caesarxuchao** [comment]
Running `:GoUpdateBinaries` inside vim fixed vim-go for me.

**Comment by thockin** [comment]
So the two proposals on the table are:

1) evict api, client-go, api_machinery, etc from the main repo entirely,
manage them at arms length, vendor back into main repo

2) keep staging (maybe under a better name) as the true home of these
packages, continue to publish from staging -> external repos, but not
consuming those repos ourselves; document better

(1) has the advantages of being obvious and fixing many of github's issues;
but the disadvantage of being more work for cross-repo changes

(2) has the advantages of being easier to integration-test and do
cross-repo changes; but the disadvantages of being somewhat alien and
requiring tooling changes (e.g. for symlinks).

Other considerations?

Chao & Daniel - at some point we talked about abusing GOPATH more, but we
settled on symlinks.  I can see the "works with go tooling" argument - are
there any others?

**Comment by sttts** [comment]
To give them a name:

(2a) keep using symlinks
(2b) get rid of staging and move everything into vendor
(2c) GOPATH=$KUBE_ROOT:$KUBE_ROOT/staging

(2a) has especially IDE issues like with code navigation and grepping
  - Gogland gets confused and doesn't understand the files are the same in vendor and staging
  - Vim-go seems to be fixable
  - git grep is fine. Go build is fine.
  - dlv: seeing crashes with kube code. I have the suspicion that links are the reason.
(2b) any issue above the psychological one that we have authorative code in vendor/ ?
(2c) I expect the same issue as with (2a). Anybody tried?

**Comment by thockin** [comment]
(2b) breaks all vendor tools (not worse than 2a, I guess)

(2a) is more obvious than editing vendor/, which makes my skin crawl

(2c) imposes on every user, or else breaks ~all go tools.

**Comment by smarterclayton** [comment]
Can you list a set of tooling changes you'd make?

**Comment by thockin** [comment]
Hypothetically, imagine that the go compiler consumed a .GOPATH file that
specified local GOPATH records.  Then we could store multiple roots in a
single repo and know that tools built on go's stdlib would work.

We can't just bury that GOPATH muckery in our make wrapper, because then
goimports and things like that won't find it.

Alternatively, imagine godep had a way to express locally-owned deps and
make the symlinks itself.

**Comment by bgrant0607** [comment]
One comment on client-go: It would make sense to me to generate client-go and the client library used by the kubernetes repo using the same (vendored) api spec and the same (vendored) tool(s), but not actually vendor client-go into the kubernetes repo, and ideally not commit the generated code into that repo at all, to avoid rebases.

**Comment by bgrant0607** [comment]
Original issue: #2742

**Comment by lavalamp** [comment]
See: https://docs.google.com/document/d/1IHTBo6Ul6TtTD87u8OVXgt-NgjgNSzE5JQKlu7nWDmw/edit#

**Comment by ttripp** [comment]
Is anybody able to provide an overall summary of where this has all gone in the last year? Its been about a year with no updates. Thanks!

**Comment by lavalamp** [comment]
More use of `staging/` in the past year, still no one with all of: ability, social capital, and inclination to actually make the external repos the source of truth and vendor back into the main repo, as was originally intended. At this point, vgo is likely to arrive before we stop using staging/; I made #63607 to discuss how we're going to manage the transition.

**Comment by jayunit100** [comment]
tim taking a look - this is still sorta relevant

**Comment by thockin** [comment]
I think we should close this - the current staging/ isn't SO bad and more or less captures the main points here.
